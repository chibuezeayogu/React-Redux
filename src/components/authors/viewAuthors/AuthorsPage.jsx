import React, { Component } from 'react';
import _ from 'lodash';
import { array } from 'prop-types';
import { connect } from 'react-redux';
import Pagination from 'rc-pagination';
import { BounceLoader } from 'react-spinners';
import { css } from 'react-emotion';
import toastr from 'toastr';


import { loadAuthors, deleteAuthor } from '../../../actions/authorActions';
import AuthorList from './AuthorList';
import formatPagination from '../../../helpers/formatPagination';

const override = css`
    display: block;
    margin: 0 auto;
    border-color: red;
`;

/**
 * @class AuthorsPage
 * @extends Component
 */
export class AuthorsPage extends Component {
	static initialState = () => ({
		isDeleting: false,
		authors: [],
	})

	state = AuthorsPage.initialState();

	static getDerivedStateFromProps(nextProps, nextState) {
		const { currentPage } = nextState;
		let Output = {};
		if (nextProps.authors.length >= 0) {
			Output = formatPagination(nextProps.authors, currentPage);
			return {
				...AuthorsPage.initialState(),
				authors: Output.state,
				pageSize: Output.pageSize,
				totalCount: Output.totalCount,
				currentPage: Output.currentPage
			};
		}

		return null;
	}

	/**
   * @memberof AuthorsPage
   * @returns {void}
   */
	redirectToAddAuthorPage = () => {
		this.props.history.push('/author');
	}

	/**
   * @memberof AuthorsPage
	 * @param {string} authorId
   * @returns {void}
   */
	deleteAuthor = (authorId) => {
		const { courses } = this.props;
		const id = courses.findIndex(course => course.authorId === authorId);
		if (id === -1) {
			this.props.deleteAuthor(authorId);
		} else {
			toastr.error('can\'t delete author with a course');
		}
	}

	renderAuthors = () => {
		const {
			authors, pageSize, totalCount, currentPage 
		} = this.state;
		return (
			<div>
				{authors && authors.length === 0 
					? <p className="lead text-center">Authors List is Empty. Add Author</p>
					: <div>
						<table className="table table-striped">
							<thead>
								<tr>
									<th scope="col">S/N</th>
									<th scope="col">First Name</th>
									<th scope="col">Last Name</th>
									<th scope="col">&nbsp;</th>
								</tr>
							</thead>
							<tbody>
								{authors.map(author => <AuthorList 
									key={author.id}
									id={author.id}
									author={author} 
									deleteAuthor={this.deleteAuthor} 
								/>
								)}
							</tbody>
						</table>
						<Pagination
							onChange={this.handlePageChange}
							current={currentPage}
							pageSize={pageSize}
							total={totalCount}
						/>
					</div>
				}
			</div>
		);
	}

	/**
   * @memberof AuthorsPage
	 * @param {object} currentPage
	 * @param {object} pageSize
   * @returns {void}
   */
	handlePageChange = (currentPage) => {
		this.setState({ currentPage });
	} 
	

	render() {
		const { isLoading } = this.props;

		return (
			<div>
				<div className="jumbotron">
					<h1 className="display-4">Authors Page</h1>
					<button type="button"
						className="btn btn-primary"
						onClick={this.redirectToAddAuthorPage}>
						Add Author
					</button>
				</div>
				{ isLoading
					? <div className="sweet-loading">
						<BounceLoader
							className={override}
							sizeUnit={'px'}
							size={150}
							color={'#123abc'}
							loading={true}
						/>
					</div> : this.renderAuthors()
				}
			</div>
		);
	}
}

AuthorsPage.propTypes = {
	authors: array.isRequired
};

const mapStateToProps = ({ authorsReducer, coursesReducer }) => ({
	authors: authorsReducer.authors,
	isLoading: authorsReducer.isLoading,
	courses: coursesReducer.courses
});

export default connect(
	mapStateToProps,
	{ 
		loadAuthors,
		deleteAuthor
	}
)(AuthorsPage);