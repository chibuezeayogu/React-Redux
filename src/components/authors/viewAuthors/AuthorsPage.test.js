import React from 'react';
import { shallow } from 'enzyme';

import { AuthorsPage } from './AuthorsPage';
import courses from '../../../__test__/__mocks__/mockCourses';
import authors from '../../../__test__/__mocks__/mockAuthors';

describe('AuthorsPage Component', () => {
	const props = {
		history: {
			push: jest.fn()
		},
		courses,
		authors,
		deleteAuthor: jest.fn().mockImplementation(() => Promise.resolve())
	};
	const wrapper = shallow(<AuthorsPage {...props}/>);
  
	it('should render correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should call redirectToAddAuthorPage method', () => {
		const spy = jest.spyOn(wrapper.instance(), 'redirectToAddAuthorPage');
		wrapper.instance().redirectToAddAuthorPage();
		expect(spy).toHaveBeenCalled();
	});
	
	it('should call deleteCourse method', () => {
		const deleteAuthorSpy = jest.spyOn(wrapper.instance(), 'deleteAuthor');
		wrapper.instance().deleteAuthor();
		expect(deleteAuthorSpy).toHaveBeenCalled();
	});
	
	it('should call handlePageChange method', () => {
		const spy = jest.spyOn(wrapper.instance(), 'handlePageChange');
		wrapper.instance().handlePageChange();
		expect(spy).toHaveBeenCalled();
	});
});
