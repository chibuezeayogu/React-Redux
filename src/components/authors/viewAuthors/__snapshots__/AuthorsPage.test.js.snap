// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorsPage Component should render correctly 1`] = `
<div>
  <div
    className="jumbotron"
  >
    <h1
      className="display-4"
    >
      Authors Page
    </h1>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Add Author
    </button>
  </div>
  <div>
    <div>
      <table
        className="table table-striped"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              S/N
            </th>
            <th
              scope="col"
            >
              First Name
            </th>
            <th
              scope="col"
            >
              Last Name
            </th>
            <th
              scope="col"
            >
               
            </th>
          </tr>
        </thead>
        <tbody>
          <AuthorList
            author={
              Object {
                "firstName": "Cory",
                "id": "cory-house",
                "lastName": "House",
              }
            }
            deleteAuthor={[Function]}
            id="cory-house"
            key="cory-house"
          />
          <AuthorList
            author={
              Object {
                "firstName": "Scott",
                "id": "scott-allen",
                "lastName": "Allen",
              }
            }
            deleteAuthor={[Function]}
            id="scott-allen"
            key="scott-allen"
          />
          <AuthorList
            author={
              Object {
                "firstName": "Dan",
                "id": "dan-wahlin",
                "lastName": "Wahlin",
              }
            }
            deleteAuthor={[Function]}
            id="dan-wahlin"
            key="dan-wahlin"
          />
        </tbody>
      </table>
      <Pagination
        className=""
        current={1}
        defaultCurrent={1}
        defaultPageSize={10}
        hideOnSinglePage={false}
        itemRender={[Function]}
        locale={
          Object {
            "items_per_page": "条/页",
            "jump_to": "跳至",
            "jump_to_confirm": "确定",
            "next_3": "向后 3 页",
            "next_5": "向后 5 页",
            "next_page": "下一页",
            "page": "页",
            "prev_3": "向前 3 页",
            "prev_5": "向前 5 页",
            "prev_page": "上一页",
          }
        }
        onChange={[Function]}
        onShowSizeChange={[Function]}
        pageSize={4}
        prefixCls="rc-pagination"
        selectComponentClass={null}
        selectPrefixCls="rc-select"
        showLessItems={false}
        showPrevNextJumpers={true}
        showQuickJumper={false}
        showSizeChanger={false}
        showTitle={true}
        style={Object {}}
        total={3}
      />
    </div>
  </div>
</div>
`;
